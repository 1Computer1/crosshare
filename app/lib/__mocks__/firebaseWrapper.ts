const firebaseTesting = jest.requireActual('@firebase/testing');
import type firebaseAdminType from 'firebase-admin';

export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const AuthProvider = null;

export const DeleteSentinal = firebaseTesting.firestore.FieldValue.delete();

export const ServerTimestamp = firebaseTesting.firestore.FieldValue.serverTimestamp();

export const TimestampClass = firebaseTesting.firestore.Timestamp;
export const AdminTimestamp = firebaseTesting.firestore.Timestamp;

let anonApp: firebase.app.App;
let anonUser: firebase.User;
export const setUpForSignInAnonymously = (app: firebase.app.App, user: firebase.User) => {
  anonApp = app;
  anonUser = user;
};

export const signInAnonymously = async () => {
  setApp(anonApp);
  return anonUser;
};
