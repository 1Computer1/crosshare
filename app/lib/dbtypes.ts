import * as t from 'io-ts';

// We import this instead of defining it here so that we can define it
// differently for the main react app and the 'functions' subdirectory
import { timestamp } from './timestamp';
export { timestamp } from './timestamp';

const CommentV = t.type({
  /** comment text */
  c: t.string,
  /** author id */
  a: t.string,
  /** author display name */
  n: t.string,
  /** author solve time in fractional seconds */
  t: t.number,
  /** author did cheat? */
  ch: t.boolean,
  /** comment publish timestamp */
  p: timestamp
});
type CommentT = t.TypeOf<typeof CommentV>;

export interface CommentWithRepliesT extends CommentT {
  /** comment id */
  i: string,
  /** replies */
  r?: Array<CommentWithRepliesT>
}

const CommentWithRepliesV: t.Type<CommentWithRepliesT> = t.recursion('CommentWithReplies', () =>
  t.intersection([
    CommentV,
    t.type({
      /** comment id */
      i: t.string,
    }),
    t.partial({
      /** replies */
      r: t.array(CommentWithRepliesV)
    })
  ])
);

export const CommentForModerationV = t.intersection([
  CommentV,
  t.type({
    /** puzzle id */
    pid: t.string,
    /** id of the comment this is a reply to */
    rt: t.union([t.string, t.null])
  })
]);
export type CommentForModerationT = t.TypeOf<typeof CommentForModerationV>;

export const CommentForModerationWithIdV = t.intersection([
  CommentForModerationV,
  t.type({
    /** comment id */
    i: t.string,
  })
]);
export type CommentForModerationWithIdT = t.TypeOf<typeof CommentForModerationWithIdV>;

const DBPuzzleMandatoryV = t.type({
  /** created at */
  ca: timestamp,
  /** author's user id */
  a: t.string,
  /** author's display name */
  n: t.string,
  /** category */
  c: t.union([t.string, t.null]),
  /** is this puzzle moderated? */
  m: t.boolean,
  /** timestamp when the puzzle goes live */
  p: t.union([timestamp, t.null]),
  /** title */
  t: t.string,
  /** grid width / columns */
  w: t.number,
  /** grid height / rows */
  h: t.number,
  /** across clue strings */
  ac: t.array(t.string),
  /** across clue display numbers */
  an: t.array(t.number),
  /** down clue strings */
  dc: t.array(t.string),
  /** down clue display numbers */
  dn: t.array(t.number),
  /** grid (solution) */
  g: t.array(t.string),
});
const DBPuzzleOptionalV = t.partial({
  /** highlighted cell indexes */
  hs: t.array(t.number),
  /** use shade instead of circle for highlight? */
  s: t.boolean,
  /** comments */
  cs: t.array(CommentWithRepliesV),
  /** is this puzzle featured (should appear on homepage) */
  f: t.boolean,
});
export const DBPuzzleV = t.intersection([DBPuzzleMandatoryV, DBPuzzleOptionalV]);
export type DBPuzzleT = t.TypeOf<typeof DBPuzzleV>;

export const TimestampedPuzzleV = downloadTimestamped(DBPuzzleV);
export type TimestampedPuzzleT = t.TypeOf<typeof TimestampedPuzzleV>;

const PlayBaseV = t.type({
  /** crossword id */
  c: t.string,
  /** updated at */
  ua: timestamp,
  /** filled in grid */
  g: t.array(t.string),
  /** play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** update iteration count per cell */
  uc: t.array(t.number),
  /** list of (checked) correct cells */
  vc: t.array(t.number),
  /** list of (checked) wrong cells */
  wc: t.array(t.number),
  /** list of cells ever marked wrong */
  we: t.array(t.number),
  /** list of revealed cells */
  rc: t.array(t.number),
  /** total play time, in fractional seconds */
  t: t.number,
  /** did "cheat"? */
  ch: t.boolean,
  /** finished the puzzle? */
  f: t.boolean,
});

export const LegacyPlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
  }),
  t.partial({
    /** puzzle title, optional for legacy plays in the db */
    n: t.string,
  })
]);
export type LegacyPlayT = t.TypeOf<typeof LegacyPlayV>;

export const PlayV = t.intersection([
  PlayBaseV,
  t.type({
    /** user id */
    u: t.string,
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayT = t.TypeOf<typeof PlayV>;

// We don't need a user id in local storage, but we do need a title
export const PlayWithoutUserV = t.intersection([
  PlayBaseV,
  t.type({
    /** puzzle title */
    n: t.string,
  })
]);
export type PlayWithoutUserT = t.TypeOf<typeof PlayWithoutUserV>;

export function downloadTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: timestamp,
    data: type,
  });
}

export function downloadOptionallyTimestamped<A>(type: t.Type<A>) {
  return t.type({
    downloadedAt: t.union([timestamp, t.null]),
    data: type,
  });
}

export function getDateString(pd: Date) {
  return pd.getUTCFullYear() + '-' + pd.getUTCMonth() + '-' + pd.getUTCDate();
}

export function prettifyDateString(dateString: string) {
  const groups = dateString.match(/^(\d+)-(\d+)-(\d+)$/);
  if (!groups) {
    throw new Error('Bad date string: ' + dateString);
  }
  return (parseInt(groups[2]) + 1) + '/' + parseInt(groups[3]) + '/' + groups[1];
}

export const PuzzleStatsV = t.type({
  /** author id, denormalized for security rules purposes. */
  a: t.string,
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** total completions without cheats */
  s: t.number,
  /** total time spent on puzzle, in fractional seconds */
  nt: t.number,
  /** total time spent by those w/o cheats */
  st: t.number,
  /** total play time of last update to each cell, in fractional seconds */
  ct: t.array(t.number),
  /** total update iteration count for each cell */
  uc: t.array(t.number),
});
export type PuzzleStatsT = t.TypeOf<typeof PuzzleStatsV>;

export const DailyStatsV = t.type({
  /** updated at */
  ua: timestamp,
  /** total completions */
  n: t.number,
  /** user ids with completions */
  u: t.array(t.string),
  /** completions by puzzleId */
  c: t.record(t.string, t.number),
  /** completions by hour (as UTC 0-23) */
  h: t.array(t.number),
});
export type DailyStatsT = t.TypeOf<typeof DailyStatsV>;

export const CronStatusV = t.type({
  ranAt: timestamp,
});
export type CronStatusT = t.TypeOf<typeof CronStatusV>;

/** created at, title */
const AuthoredPuzzleV = t.tuple([timestamp, t.string]);
export type AuthoredPuzzleT = t.TypeOf<typeof AuthoredPuzzleV>;

/** keys are puzzle ids */
export const AuthoredPuzzlesV = t.record(t.string, AuthoredPuzzleV);

/** date string -> puzzle id */
export const CategoryIndexV = t.record(t.string, t.string);
export type CategoryIndexT = t.TypeOf<typeof CategoryIndexV>;
