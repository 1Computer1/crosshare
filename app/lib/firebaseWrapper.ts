// On client side we include firebase via script tags.
// The 'externals' field in package.json tells it not to bundle these.
import * as firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/performance';
import 'firebase/auth';
import 'firebase/storage';

import type * as firebaseAdminType from 'firebase-admin';

const firebaseConfig = {
  apiKey: 'AIzaSyBrmmBf91peVT5T_Z7N3z9oizsPH5u2pUc',
  authDomain: 'auth.crosshare.org',
  databaseURL: 'https://mdcrosshare.firebaseio.com',
  projectId: 'mdcrosshare',
  storageBucket: 'mdcrosshare.appspot.com',
  messagingSenderId: '603173482014',
  appId: '1:603173482014:web:98d7d820731b7c5eaa080f',
  measurementId: 'G-LTLN7Z4XBS'
};

// Initialize Firebase
export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;
export let AdminTimestamp: typeof firebaseAdminType.firestore.Timestamp;

if (typeof window === 'undefined') {
  const firebaseAdmin: typeof firebaseAdminType = require('firebase-admin'); // eslint-disable-line @typescript-eslint/no-var-requires
  AdminTimestamp = firebaseAdmin.firestore.Timestamp;
  if (firebaseAdmin.apps.length && firebaseAdmin.apps[0]) {
    AdminApp = firebaseAdmin.apps[0];
  } else {
    AdminApp = firebaseAdmin.initializeApp({
      ...firebaseConfig,
      credential: firebaseAdmin.credential.applicationDefault(),
    });
  }
}

if (firebase.apps.length) {
  App = firebase.apps[0];
} else {
  App = firebase.initializeApp(firebaseConfig);
  if (typeof window !== 'undefined') {
    if (process.env.NODE_ENV === 'production') {
      try {
        App.performance();
      } catch {
        console.error('failed to load firebase.performance');
      }
    }
  }
  if (process.env.NEXT_PUBLIC_USE_FIREBASE_EMULATOR) {
    const db = App.firestore();
    db.settings({
      host: 'localhost:8080',
      ssl: false
    });
  }
}

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const setUpForSignInAnonymously = (_app: firebase.app.App, _user: firebase.User) => {
  throw new Error('For testing only');
};

export const signInAnonymously = async () => {
  const userCredential = await firebase.auth().signInAnonymously();
  if (!userCredential.user) {
    throw new Error('Logged in anonymously but no user in result');
  }
  return userCredential.user;
};

export const setUserMap = (_map: Record<string, firebase.User>) => {/* noop */ };
export const getUser = (userId: string) => AdminApp.auth().getUser(userId);

export const AuthProvider = new firebase.auth.GoogleAuthProvider();

export const DeleteSentinal = firebase.firestore.FieldValue.delete();

export const ServerTimestamp = firebase.firestore.FieldValue.serverTimestamp();

export const FieldValue = firebase.firestore.FieldValue;

export let TimestampClass = firebase.firestore.Timestamp;
export function setTimestampClass(cls: any) { //eslint-disable-line @typescript-eslint/no-explicit-any
  TimestampClass = cls;
}
export type TimestampType = firebase.firestore.Timestamp;
