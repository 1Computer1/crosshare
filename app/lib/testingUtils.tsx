import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
