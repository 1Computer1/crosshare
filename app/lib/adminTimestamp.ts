/* this is the firebase timestamp type for the main react project! */

import * as t from 'io-ts';
import { either } from 'fp-ts/lib/Either';

import { AdminTimestamp } from './firebaseWrapper';
import type firebaseAdminType from 'firebase-admin';

const isFirestoreTimestamp = (u: unknown): u is firebaseAdminType.firestore.Timestamp =>
  u ? u instanceof AdminTimestamp : false;

const validateTimestamp: t.Validate<unknown, firebaseAdminType.firestore.Timestamp> = (i, c) => {
  if (isFirestoreTimestamp(i)) {
    return t.success(i);
  }
  return either.chain(
    t.type({ seconds: t.number, nanoseconds: t.number }).validate(i, c),
    obj => t.success(new AdminTimestamp(obj.seconds, obj.nanoseconds))
  );
};

export const adminTimestamp = new t.Type<firebaseAdminType.firestore.Timestamp>(
  'AdminTimestamp',
  isFirestoreTimestamp,
  validateTimestamp,
  t.identity,
);
