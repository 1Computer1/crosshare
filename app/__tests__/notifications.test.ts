import * as firebaseTesting from '@firebase/testing';

import { getMockedPuzzle } from '../lib/testingUtils';
import { notificationsForPuzzleChange } from '../lib/notifications';
import { CommentWithRepliesT } from '../lib/dbtypes';
import { TimestampClass, setAdminApp, setUserMap } from '../lib/firebaseWrapper';
import { queueEmails } from '../lib/serverOnly';
import type firebaseAdminType from 'firebase-admin';
import add from 'date-fns/add';

jest.mock('../lib/firebaseWrapper');

const basePuzzle = getMockedPuzzle({ cs: undefined });

function getComment(fields?: Partial<CommentWithRepliesT>): CommentWithRepliesT {
  return {
    ...{
      c: 'A couple of two-worders today which I don\'t love, but I hope you all got it anyway!',
      i: 'LwgoVx0BAskM4wVJyoLj',
      t: 36.009,
      p: TimestampClass.now(),
      a: 'fSEwJorvqOMK5UhNMHa4mu48izl1',
      n: 'Mike D',
      ch: false,
    },
    ...fields
  };
}

test('shouldnt notify at all if comment is on own puzzle', () => {
  const puzzleWithComments = {
    ...basePuzzle,
    cs: [getComment({ a: basePuzzle.a })]
  };
  const notifications = notificationsForPuzzleChange(basePuzzle, puzzleWithComments, 'puzzle-id-here');
  expect(notifications.length).toEqual(0);
});

const projectId = 'notificationstests';

test('security rules for updating notifications', async () => {
  const puzzleWithComments = {
    ...basePuzzle,
    cs: [getComment({ a: 'dummy-author-id' })]
  };
  const notifications = notificationsForPuzzleChange(basePuzzle, puzzleWithComments, 'puzzle-id-here');
  expect(notifications.length).toEqual(1);

  await firebaseTesting.clearFirestoreData({ projectId });
  const adminApp = firebaseTesting.initializeAdminApp({ projectId }) as firebase.app.App;
  await adminApp.firestore().collection('n').doc(notifications[0].id).set(notifications[0]);

  const ownerApp = firebaseTesting.initializeTestApp({
    projectId,
    auth: {
      uid: notifications[0].u,
      admin: false,
      firebase: {
        sign_in_provider: 'google.com',
      },
    },
  });
  const otherApp = firebaseTesting.initializeTestApp({
    projectId,
    auth: {
      uid: 'randouserid',
      admin: false,
      firebase: {
        sign_in_provider: 'google.com',
      },
    },
  });

  // other user can't change read status
  await firebaseTesting.assertFails(
    otherApp.firestore().collection('n').doc(notifications[0].id).update({ r: true })
  );
  // owner can't change anything other than read status
  await firebaseTesting.assertFails(
    ownerApp.firestore().collection('n').doc(notifications[0].id).update({ c: 'some new comment id' })
  );
  // owner can change read status
  await firebaseTesting.assertSucceeds(
    ownerApp.firestore().collection('n').doc(notifications[0].id).update({ r: true })
  );

  adminApp.delete();
  ownerApp.delete();
  otherApp.delete();
});

test('should notify for a new comment by somebody else', () => {
  const puzzleWithComments = {
    ...basePuzzle,
    cs: [getComment({ a: 'dummy-author-id' })]
  };
  const notifications = notificationsForPuzzleChange(basePuzzle, puzzleWithComments, 'puzzle-id-here');
  expect(notifications.length).toEqual(1);
  notifications.forEach(n => { delete n.t; });
  expect(notifications).toMatchSnapshot();
});

test('should notify for multiple comments by somebody else', () => {
  const puzzleWithComments = {
    ...basePuzzle,
    cs: [
      getComment({ a: basePuzzle.a, i: 'foo' }),
      getComment({ a: 'dummy-author-id', i: 'bar', n: 'Jim' }),
      getComment({ a: 'another-author', i: 'bam', n: 'Tom' }),
    ]
  };
  const notifications = notificationsForPuzzleChange(basePuzzle, puzzleWithComments, 'puzzle-id-here');
  expect(notifications.length).toEqual(2);
  notifications.forEach(n => { delete n.t; });
  expect(notifications).toMatchSnapshot();
});

test('should notify for a reply to own comment on own puzzle', () => {
  const puzzleWithOwn = {
    ...basePuzzle,
    cs: [getComment({ a: basePuzzle.a })]
  };
  const puzzleWithComments = {
    ...basePuzzle,
    cs: [getComment({ a: basePuzzle.a, r: [getComment({ a: 'dummy-author-id', i: 'bar', n: 'Jim' })] })]
  };
  const notifications = notificationsForPuzzleChange(puzzleWithOwn, puzzleWithComments, 'puzzle-id-here');
  expect(notifications.length).toEqual(1);
  notifications.forEach(n => { delete n.t; });
  expect(notifications).toMatchSnapshot();
});

test('should notify comment author only when puzzle author replies', () => {
  const puzzleWithComment = {
    ...basePuzzle,
    cs: [getComment({ a: 'dummy-author-id' })]
  };
  const authorReplies = {
    ...basePuzzle,
    cs: [getComment({ a: 'dummy-author-id', r: [getComment({ a: basePuzzle.a, i: 'baz' })] })]
  };
  const notifications = notificationsForPuzzleChange(puzzleWithComment, authorReplies, 'puzzle-id-here');
  expect(notifications.length).toEqual(1);
  notifications.forEach(n => { delete n.t; });
  expect(notifications).toMatchSnapshot();
});

test('should notify comment author and puzzle author when third party replies', () => {
  const puzzleWithComment = {
    ...basePuzzle,
    cs: [getComment({ a: 'dummy-author-id' })]
  };
  const authorReplies = {
    ...basePuzzle,
    cs: [getComment({ a: 'dummy-author-id', r: [getComment({ a: 'rando', i: 'baz' })] })]
  };
  const notifications = notificationsForPuzzleChange(puzzleWithComment, authorReplies, 'puzzle-id-here');
  expect(notifications.length).toEqual(2);
  notifications.forEach(n => { delete n.t; });
  expect(notifications).toMatchSnapshot();
});

test('should handle a combination of multiple new comments and nested replies', () => {
  const startingPoint = {
    ...basePuzzle,
    cs: [getComment({ r: [getComment({ a: 'rando', i: 'baz' })] })]
  };
  const withReplies = {
    ...basePuzzle,
    cs: [
      getComment({ a: 'blaster', i: 'bam', n: 'BLAST' }),
      getComment({
        r: [
          getComment({
            a: 'rando', i: 'baz', r: [
              getComment({ i: 'whamo' }),
              getComment({ a: 'blaster', i: 'test' })
            ]
          }),
          getComment({ a: 'another-rando', i: 'foobar' })
        ]
      })
    ]
  };
  const notifications = notificationsForPuzzleChange(startingPoint, withReplies, 'puzzle-id-here');
  expect(notifications.length).toEqual(5);
  notifications.forEach(n => { delete n.t; });
  expect(notifications).toMatchSnapshot();
});

describe('email queueing', () => {
  let adminApp: firebaseAdminType.app.App;
  const realDate = Date;

  beforeEach(async () => {
    const startingPoint = {
      ...basePuzzle,
      cs: [getComment({ r: [getComment({ a: 'rando', i: 'baz' })] })]
    };
    const withReplies = {
      ...basePuzzle,
      cs: [
        getComment({ a: 'blaster', i: 'bam', n: 'BLAST' }),
        getComment({
          r: [
            getComment({
              a: 'rando', i: 'baz', r: [
                getComment({ i: 'whamo' }),
                getComment({ a: 'blaster', i: 'test' })
              ]
            }),
            getComment({ a: 'another-rando', i: 'foobar' })
          ]
        })
      ]
    };
    const notifications = notificationsForPuzzleChange(startingPoint, withReplies, 'puzzle-id-here');
    expect(notifications.length).toEqual(5);
    await firebaseTesting.clearFirestoreData({ projectId });
    adminApp = firebaseTesting.initializeAdminApp({ projectId }) as unknown as firebaseAdminType.app.App;
    setAdminApp(adminApp);
    for (const notification of notifications) {
      await adminApp.firestore().collection('n').doc(notification.id).set(notification);
    }
    setUserMap({ rando: { email: 'rando@example.com' } as firebase.User, [basePuzzle.a]: { email: 'mike@example.com' } as firebase.User });
  });

  afterEach(async () => {
    await adminApp.delete();
  });

  async function withMockedDate(work: () => Promise<void>) {
    const mockDate = add(new Date(), { hours: 2 });
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).Date = class extends realDate {
      constructor() {
        super();
        return mockDate;
      }
    };
    await work();

    global.Date = realDate;
  }

  test('nothing sends until an hour after comments are posted', async () => {
    await queueEmails();
    const mail = await adminApp.firestore().collection('mail').get();
    expect(mail.size).toEqual(0);
  });

  test('emails send after an hour, but dont double send', async () => {
    return withMockedDate(async () => {
      await queueEmails();
      // we can do it again immediately and we shouldn't double send due to marking as read
      await queueEmails();

      const mail2 = await adminApp.firestore().collection('mail').get();
      expect(mail2.size).toEqual(2);
      expect(mail2.docs.map(d => d.data()).sort((a, b) => a.to[0].localeCompare(b.to[0]))).toMatchSnapshot();
    });
  });

  test('dont send if unsubscribed to all', async () => {
    return withMockedDate(async () => {
      await adminApp.firestore().collection('prefs').doc('rando').set({ unsubs: ['all'] }, { merge: true });

      await queueEmails();

      // We should still have marked everything as read
      expect((await adminApp.firestore().collection('n').where('r', '==', false).get()).size).toEqual(0);

      const mail2 = await adminApp.firestore().collection('mail').get();
      expect(mail2.size).toEqual(1);
      expect(mail2.docs[0].data()).toMatchSnapshot();
    });
  });

  test('dont send if unsubscribed to comments and thats all we have', async () => {
    return withMockedDate(async () => {
      await adminApp.firestore().collection('prefs').doc(basePuzzle.a).set({ unsubs: ['comments'] }, { merge: true });

      await queueEmails();

      // We should still have marked everything as read
      expect((await adminApp.firestore().collection('n').where('r', '==', false).get()).size).toEqual(0);

      const mail2 = await adminApp.firestore().collection('mail').get();
      expect(mail2.size).toEqual(1);
      expect(mail2.docs[0].data()).toMatchSnapshot();
    });
  });

  test('emails work w/ multiple puzzles and w/ html in title', async () => {
    const puzzleWithComment = getMockedPuzzle({
      t: 'Our Second <span> Title',
      cs: [getComment({ a: 'rando' })]
    });
    const withReplies = {
      ...puzzleWithComment,
      cs: [getComment({ a: 'rando', r: [getComment({ a: 'anotherAuthor', n: 'Commenter <div> Foo', i: 'baz' })] })]
    };
    const notifications = notificationsForPuzzleChange(puzzleWithComment, withReplies, 'second-puzzle');
    expect(notifications.length).toEqual(2);
    for (const notification of notifications) {
      await adminApp.firestore().collection('n').doc(notification.id).set(notification);
    }

    return withMockedDate(async () => {

      await queueEmails();

      const mail2 = await adminApp.firestore().collection('mail').get();
      expect(mail2.size).toEqual(2);
      expect(mail2.docs.map(d => d.data()).sort((a, b) => a.to[0].localeCompare(b.to[0]))).toMatchSnapshot();
    });
  });

  test('emails work w/ multiple categories of comments', async () => {
    const puzzleWithComment = getMockedPuzzle({
      a: 'rando',
      t: 'Our Second Title',
      cs: [getComment({ a: basePuzzle.a })]
    });
    const withReplies = {
      ...puzzleWithComment,
      cs: [getComment({ a: basePuzzle.a, r: [getComment({ a: 'anotherAuthor', n: 'Foo Bar', i: 'baz' })] })]
    };
    const notifications = notificationsForPuzzleChange(puzzleWithComment, withReplies, 'second-puzzle');
    expect(notifications.length).toEqual(2);
    for (const notification of notifications) {
      await adminApp.firestore().collection('n').doc(notification.id).set(notification);
    }

    return withMockedDate(async () => {

      await queueEmails();

      const mail2 = await adminApp.firestore().collection('mail').get();
      expect(mail2.size).toEqual(2);
      expect(mail2.docs.map(d => d.data()).sort((a, b) => a.to[0].localeCompare(b.to[0]))).toMatchSnapshot();
    });
  });
});
