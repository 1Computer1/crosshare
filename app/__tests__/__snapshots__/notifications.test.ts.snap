// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email queuing 1`] = `Array []`;

exports[`email queuing 2`] = `
Array [
  Object {
    "message": Object {
      "html": "<h3>Comments on your puzzles:</h3><ul><li>BLAST and Mike D commented on <a href=\\"https://crosshare.org/crosswords/puzzle-id-here#utm_source=crosshare&amp;utm_medium=email&amp;utm_campaign=notifications\\">Raises, as young</a></li></ul><h3>Replies to your comments:</h3><ul><li>Mike D replied to your comment(s) on <a href=\\"https://crosshare.org/crosswords/puzzle-id-here#utm_source=crosshare&amp;utm_medium=email&amp;utm_campaign=notifications\\">Raises, as young</a></li></ul>",
      "subject": "New comments on Raises, as young",
      "text": "### Comments on your puzzles:

* BLAST and Mike D commented on [Raises, as young](https://crosshare.org/crosswords/puzzle-id-here#utm_source=crosshare&utm_medium=email&utm_campaign=notifications)


### Replies to your comments:

* Mike D replied to your comment(s) on [Raises, as young](https://crosshare.org/crosswords/puzzle-id-here#utm_source=crosshare&utm_medium=email&utm_campaign=notifications)


",
    },
    "to": Array [
      "mike@example.com",
    ],
  },
  Object {
    "message": Object {
      "html": "<h3>Replies to your comments:</h3><ul><li>Mike D and Mike D replied to your comment(s) on <a href=\\"https://crosshare.org/crosswords/puzzle-id-here#utm_source=crosshare&amp;utm_medium=email&amp;utm_campaign=notifications\\">Raises, as young</a></li></ul>",
      "subject": "Replies to your comment on  and ",
      "text": "### Replies to your comments:

* Mike D and Mike D replied to your comment(s) on [Raises, as young](https://crosshare.org/crosswords/puzzle-id-here#utm_source=crosshare&utm_medium=email&utm_campaign=notifications)


",
    },
    "to": Array [
      "rando@example.com",
    ],
  },
]
`;

exports[`should handle a combination of multiple new comments and nested replies 1`] = `
Array [
  Object {
    "c": "bam",
    "cn": "BLAST",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-comment-bam",
    "k": "comment",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
  Object {
    "c": "whamo",
    "cn": "Mike D",
    "id": "rando-reply-whamo",
    "k": "reply",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "rando",
  },
  Object {
    "c": "test",
    "cn": "Mike D",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-comment-test",
    "k": "comment",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
  Object {
    "c": "test",
    "cn": "Mike D",
    "id": "rando-reply-test",
    "k": "reply",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "rando",
  },
  Object {
    "c": "foobar",
    "cn": "Mike D",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-reply-foobar",
    "k": "reply",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
]
`;

exports[`should notify comment author and puzzle author when third party replies 1`] = `
Array [
  Object {
    "c": "baz",
    "cn": "Mike D",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-comment-baz",
    "k": "comment",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
  Object {
    "c": "baz",
    "cn": "Mike D",
    "id": "dummy-author-id-reply-baz",
    "k": "reply",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "dummy-author-id",
  },
]
`;

exports[`should notify comment author only when puzzle author replies 1`] = `
Array [
  Object {
    "c": "baz",
    "cn": "Mike D",
    "id": "dummy-author-id-reply-baz",
    "k": "reply",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "dummy-author-id",
  },
]
`;

exports[`should notify for a new comment by somebody else 1`] = `
Array [
  Object {
    "c": "LwgoVx0BAskM4wVJyoLj",
    "cn": "Mike D",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-comment-LwgoVx0BAskM4wVJyoLj",
    "k": "comment",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
]
`;

exports[`should notify for a reply to own comment on own puzzle 1`] = `
Array [
  Object {
    "c": "bar",
    "cn": "Jim",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-reply-bar",
    "k": "reply",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
]
`;

exports[`should notify for multiple comments by somebody else 1`] = `
Array [
  Object {
    "c": "bar",
    "cn": "Jim",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-comment-bar",
    "k": "comment",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
  Object {
    "c": "bam",
    "cn": "Tom",
    "id": "fSEwJorvqOMK5UhNMHa4mu48izl1-comment-bam",
    "k": "comment",
    "p": "puzzle-id-here",
    "pn": "Raises, as young",
    "r": false,
    "u": "fSEwJorvqOMK5UhNMHa4mu48izl1",
  },
]
`;
